lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -58,13 +94,63 @@ FModuleOp LowerLayersPass::buildNewModule(OpBuilder &builder, Location location,
   return newModule;
 }
 
-/// Process a module to remove any layer blocks it has.
-void LowerLayersPass::runOnModule(FModuleOp moduleOp) {
+void LowerLayersPass::removeLayersFromValue(Value value) {
+  auto type = dyn_cast<RefType>(value.getType());
+  if (!type || !type.getLayer())
+    return;
+  value.setType(type.removeLayer());
+}
+
+void LowerLayersPass::removeLayersFromPorts(FModuleLike moduleLike) {
+  if (auto moduleOp = dyn_cast<FModuleOp>(moduleLike.getOperation())) {
+    for (auto arg : moduleOp.getBodyBlock()->getArguments())
+      removeLayersFromValue(arg);
+  }
+
+  auto oldTypeAttrs = moduleLike.getPortTypesAttr();
+  SmallVector<Attribute> newTypeAttrs;
+  newTypeAttrs.reserve(oldTypeAttrs.size());
+  for (auto typeAttr : oldTypeAttrs.getAsRange<TypeAttr>()) {
+    if (auto refType = dyn_cast<RefType>(typeAttr.getValue()))
+      if (refType.getLayer())
+        typeAttr = TypeAttr::get(refType.removeLayer());
+    newTypeAttrs.push_back(typeAttr);
+  }
+  moduleLike->setAttr(FModuleLike::getPortTypesAttrName(),
+                      ArrayAttr::get(moduleLike.getContext(), newTypeAttrs));
```

Comment:
I think this will rewrite the attributes even if there is no change. That may be inefficient if this is causing the entire attribute dictionary to be recreated and interned. This would be better to have an exit that doesn't rewrite if there is no work to do.

Reply:
Updated.